packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx

```diff
@@ -0,0 +1,145 @@
+import * as React from 'react';
+import { unstable_composeClasses as composeClasses } from '@mui/material';
+import {
+  useGridLogger,
+  GridEvents,
+  useGridApiEventHandler,
+  GridEventListener,
+  getDataGridUtilityClass,
+  useGridSelector,
+  gridSortModelSelector,
+  gridRowTreeDepthSelector,
+  useGridApiOptionHandler,
+  GridRowId,
+} from '@mui/x-data-grid';
+import { GridRowOrderChangeParams } from '@mui/x-data-grid-pro/models';
+import { GridApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+
+type OwnerState = { classes: DataGridProProcessedProps['classes'] };
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    rowDragging: ['row--dragging'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+/**
+ * Only available in DataGridPro
+ * @requires useGridRows (method)
+ */
+export const useGridRowReorder = (
+  apiRef: React.MutableRefObject<GridApiPro>,
+  props: Pick<DataGridProProcessedProps, 'rowReordering' | 'onRowOrderChange' | 'classes'>,
+): void => {
+  const logger = useGridLogger(apiRef, 'useGridRowReorder');
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const treeDepth = useGridSelector(apiRef, gridRowTreeDepthSelector);
+  const dragRowNode = React.useRef<HTMLElement | null>(null);
+  const originRowIndex = React.useRef<number | null>(null);
+  const removeDnDStylesTimeout = React.useRef<any>();
+  const ownerState = { classes: props.classes };
+  const classes = useUtilityClasses(ownerState);
+  const [dragRowId, setDragRowId] = React.useState<GridRowId>('');
+
+  React.useEffect(() => {
+    return () => {
+      clearTimeout(removeDnDStylesTimeout.current);
+    };
+  }, []);
+
+  // TODO: remove sortModel check once row reorder is sorting compatible
+  // remove treeDepth once row reorder is tree compatible
+  const isRowReorderDisabled = React.useMemo((): boolean => {
+    return !props.rowReordering || !!sortModel.length || treeDepth !== 1;
+  }, [props.rowReordering, sortModel, treeDepth]);
+
+  const handleDragStart = React.useCallback<GridEventListener<GridEvents.rowDragStart>>(
+    (params, event) => {
+      if (isRowReorderDisabled) {
+        return;
+      }
+
+      logger.debug(`Start dragging row ${params.id}`);
+      // Prevent drag events propagation.
+      // For more information check here https://github.com/mui/mui-x/issues/2680.
+      event.stopPropagation();
+
+      dragRowNode.current = event.currentTarget;
+      dragRowNode.current.classList.add(classes.rowDragging);
+
+      setDragRowId(params.id);
+
+      removeDnDStylesTimeout.current = setTimeout(() => {
+        dragRowNode.current!.classList.remove(classes.rowDragging);
```

Comment:
Why is it called on drag start?

Reply:
It's a "hack", same is done for the column reorder. Basically, we want to add a class to the DOM element that is about to be dragged before the native dragging API takes over and the element is out of the DOM flow, but because there is only 1 element we want to preserve the original styling while dragging.